BaselineValidator.validateAndReplace() fails with UnsupportedOperationException under Maven 3.8.2

When using Maven 3.8.2 Tycho's `P2MetadataMojo` or `P2MetadataDefaultMojo` in version 2.4.0 (it is very likely the same for Tycho 2.5.0-SNAPSHOT because the class did not change) `BaselineValidator.validateAndReplace()` fails by throwing an `UnsupportedOperationException` when it attempts to remove an element from the list returned by `project.getAttachedArtifacts()` because the corresponding artifact is not present in the baseline.

As of Maven 3.8.2 `MavenProject.getAttachedArtifacts()` returns an unmodifiable List.
This was changed as part of the fix of [MNG-5868](https://issues.apache.org/jira/browse/MNG-5868) which was merged into the Maven master https://github.com/apache/maven/pull/358 and back-ported into the 3.8. stream as commit 66a43813770cd7dee9eb1ed0372e3f784ba0c541 (did not found the PR).

Unfortunately I did not found an alternative to remove attached artifacts from a MavenProjects.

> 
> 
> You can use a mirror: https://downloads.apache.org/maven/maven-3/3.8.2/binaries/
> I have build tycho +itests with maven 3.8.2 and got no test-failures so currently this case is not covered by itests.

Thanks for the link.
When I build m2e locally with Maven 3.8.2 the build (mvn clean verify) fails for the Default Build Lifecycle Mapping Metadata Plug-in:
```
[INFO] -----< org.eclipse.m2e:org.eclipse.m2e.lifecyclemapping.defaults >------
[INFO] Building [bundle] Default Build Lifecycle Mapping Metadata 1.17.1-SNAPSHOT [13/41]
[INFO] ---------------------------[ eclipse-plugin ]---------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.4.1:clean (default-clean) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- tycho-packaging-plugin:2.4.0:build-qualifier (default-build-qualifier) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO] The project's OSGi version is 1.17.1.20210115-1536
[INFO]
[INFO] --- tycho-packaging-plugin:2.4.0:validate-id (default-validate-id) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- tycho-packaging-plugin:2.4.0:validate-version (default-validate-version) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- target-platform-configuration:2.4.0:target-platform (default-target-platform) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- jacoco-maven-plugin:0.8.5:prepare-agent (default) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO] argLine set to -javaagent:C:\\Users\\Hannes\\.m2\\repository\\org\\jacoco\\org.jacoco.agent\\0.8.5\\org.jacoco.agent-0.8.5-runtime.jar=destfile=C:\\dev\\git\\eclipse.m2e-core\\target\\jacoco.exec,append=true,includes=org.eclipse.m2e*:org.apache.maven*:org.eclipse.aether*
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\src\main\resources
[INFO]
[INFO] --- tycho-compiler-plugin:2.4.0:compile (default-compile) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- maven-resources-plugin:2.5:testResources (default-testResources) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\src\test\resources
[INFO]
[INFO] --- tycho-compiler-plugin:2.4.0:testCompile (default-testCompile) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO]
[INFO] --- maven-surefire-plugin:2.8:test (default-test) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO] No tests to run.
[INFO] Surefire report directory: C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\target\surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
There are no tests to run.

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO]
[INFO] --- tycho-source-plugin:2.4.0:plugin-source (plugin-source) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO] Building jar: C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\target\org.eclipse.m2e.lifecyclemapping.defaults-1.17.1-SNAPSHOT-sources.jar
[INFO]
[INFO] --- tycho-packaging-plugin:2.4.0:package-plugin (default-package-plugin) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[INFO] Building jar: C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\target\org.eclipse.m2e.lifecyclemapping.defaults-1.17.1-SNAPSHOT.jar
[INFO]
[INFO] --- tycho-p2-plugin:2.4.0:p2-metadata-default (default-p2-metadata-default) @ org.eclipse.m2e.lifecyclemapping.defaults ---
[WARNING] MavenProject: org.eclipse.m2e:org.eclipse.m2e.lifecyclemapping.defaults:1.17.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\org.eclipse.m2e.lifecyclemapping.defaults\.polyglot.META-INF: baseline and build artifacts have same version but different contents
   no-classifier: different
      META-INF/ECLIPSE_.RSA: present in baseline only
      META-INF/ECLIPSE_.SF: present in baseline only
      META-INF/MANIFEST.MF: different
         Bundle-License: not present in baseline version
      about.html: different
      lifecycle-mapping-metadata.xml: different
   classifier-sources: not present in baseline

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Maven Integration for Eclipse (parent) 1.16.0-SNAPSHOT SUCCESS [  0.249 s]
[INFO] [target] m2e-target-platform 1.16.0-SNAPSHOT ....... SUCCESS [  0.193 s]
[INFO] [bundle] Maven Project Model Edit Bundle 1.18.2-SNAPSHOT SUCCESS [ 22.695 s]
[INFO] Maven Integration for Eclipse Core Plug-in 1.18.2-SNAPSHOT SUCCESS [ 15.122 s]
[INFO] [bundle] Maven Integration for Eclipse 1.18.2-SNAPSHOT SUCCESS [ 11.787 s]
[INFO] [bundle] Maven Integration for Eclipse Launching 1.17.4-SNAPSHOT SUCCESS [  3.474 s]
[INFO] [bundle] Maven Integration for Eclipse JDT 1.18.2-SNAPSHOT SUCCESS [  5.206 s]
[INFO] [bundle] Maven Integration for Eclipse JDT UI 1.17.2-SNAPSHOT SUCCESS [  2.410 s]
[INFO] [bundle] Maven Integration for Eclipse Refactoring 1.17.3-SNAPSHOT SUCCESS [  5.093 s]
[INFO] [bundle] m2e Marketplace 1.18.1-SNAPSHOT ........... SUCCESS [  4.738 s]
[INFO] [bundle] Maven Integration for Eclipse (Editors) 1.18.2-SNAPSHOT SUCCESS [ 12.023 s]
[INFO] [bundle] Maven POM XML Editor 1.18.2-SNAPSHOT ...... SUCCESS [  4.875 s]
[INFO] [bundle] Default Build Lifecycle Mapping Metadata 1.17.1-SNAPSHOT FAILURE [  1.462 s]
[INFO] [bundle] SCM Maven Integration for Eclipse 1.17.2-SNAPSHOT SKIPPED
[INFO] [bundle] Maven Profiles Management 1.17.1-SNAPSHOT . SKIPPED
[INFO] [bundle] Maven Profiles Management UI 1.17.3-SNAPSHOT SKIPPED
[INFO] [bundle] m2e extension for import framework 1.17.3-SNAPSHOT SKIPPED
[INFO] [bundle] Source Lookup Core 1.17.4-SNAPSHOT ........ SKIPPED
[INFO] [bundle] Binary project Core 1.17.4-SNAPSHOT ....... SKIPPED
[INFO] [bundle] Binary project UI 1.18.1-SNAPSHOT ......... SKIPPED
[INFO] [bundle] Source Lookup UI 1.17.3-SNAPSHOT .......... SKIPPED
[INFO] org.eclipse.m2e.editor.lemminx 1.18.2-SNAPSHOT ..... SKIPPED
[INFO] [bundle] M2E PDE Integration 1.18.2-SNAPSHOT ....... SKIPPED
[INFO] [bundle] PDE Integration UI 1.18.2-SNAPSHOT ........ SKIPPED
[INFO] [bundle] M2E Testing Helpers 1.17.4-SNAPSHOT ....... SKIPPED
[INFO] [bundle] m2e logback appender 1.17.1-SNAPSHOT ...... SKIPPED
[INFO] [bundle] m2e logback configuration 1.16.3-SNAPSHOT . SKIPPED
[INFO] [feature] POM Editor (Legacy) using WebTools SourceEditing UI 1.18.2-SNAPSHOT SKIPPED
[INFO] [feature] m2e - Maven Integration for Eclipse (includes Incubating components) 1.18.2-SNAPSHOT SKIPPED
[INFO] [feature] POM Editor using LemMinX language server (includes Incubating components) 1.18.2-SNAPSHOT SKIPPED
[INFO] [feature] m2e - Extensions Development Support (Optional) 1.18.2-SNAPSHOT SKIPPED
[INFO] [feature] m2e - slf4j over logback logging (Optional) 1.17.2-SNAPSHOT SKIPPED
[INFO] [feature] m2e PDE - Maven Integration for Eclipse Plugin Development 1.18.2-SNAPSHOT SKIPPED
[INFO] Maven Integration for Eclipse Update Site 1.16.0-SNAPSHOT SKIPPED
[INFO] org.eclipse.m2e.core.tests 1.16.1-SNAPSHOT ......... SKIPPED
[INFO] org.eclipse.m2e.core.ui.tests 1.18.1-SNAPSHOT ...... SKIPPED
[INFO] Tests for Maven Integration for Eclipse Importer framework 1.16.1-SNAPSHOT SKIPPED
[INFO] [test-bundle] Bundle-Name 1.16.0-SNAPSHOT .......... SKIPPED
[INFO] [test-bundle] Test for pom edition in generic editor 1.16.0-SNAPSHOT SKIPPED
[INFO] [test-bundle] org.eclipse.m2e.jdt.tests 1.16.2-SNAPSHOT SKIPPED
[INFO] [test-bundle] Test for pom edition in WTP-based editor 1.16.0-SNAPSHOT SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:18 min
[INFO] Finished at: 2021-09-22T08:55:00+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-plugin:2.4.0:p2-metadata-default (default-p2-metadata-default) on project org.eclipse.m2e.lifecyclemapping.defaults: Execution default-p2-metadata-default of goal org.eclipse.tycho:tycho-p2-plugin:2.4.0:p2-metadata-default failed.: UnsupportedOperationException -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.eclipse.m2e.lifecyclemapping.defaults
```

--- a/tycho-p2/tycho-p2-plugin/src/main/java/org/eclipse/tycho/plugins/p2/BaselineValidator.java
+++ b/tycho-p2/tycho-p2-plugin/src/main/java/org/eclipse/tycho/plugins/p2/BaselineValidator.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012 Sonatype Inc. and others.
+ * Copyright (c) 2012, 2021 Sonatype Inc. and others.
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -9,6 +9,7 @@
  *
  * Contributors:
  *    Sonatype Inc. - initial API and implementation
+ *    Christoph Läubrich - BaselineValidator.validateAndReplace() fails with UnsupportedOperationException under Maven 3.8.2
  *******************************************************************************/
 package org.eclipse.tycho.plugins.p2;
 
@@ -20,6 +21,7 @@ import static org.eclipse.tycho.plugins.p2.BaselineReplace.none;
 
 import java.io.File;
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -27,6 +29,7 @@ import java.util.ListIterator;
 import java.util.Map;
 import java.util.Map.Entry;
 
+import org.apache.commons.lang3.reflect.MethodUtils;
 import org.apache.maven.artifact.Artifact;
 import org.apache.maven.plugin.MojoExecution;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -145,8 +148,21 @@ public class BaselineValidator {
                             List<Artifact> attachedArtifacts = project.getAttachedArtifacts();
                             ListIterator<Artifact> iterator = attachedArtifacts.listIterator();
                             while (iterator.hasNext()) {
-                                if (classifier.equals(iterator.next().getClassifier())) {
-                                    iterator.remove();
+                                Artifact next = iterator.next();
+                                if (classifier.equals(next.getClassifier())) {
+                                    try {
+                                        iterator.remove();
+                                    } catch (UnsupportedOperationException e) {
+                                        ArrayList<Artifact> list = new ArrayList<>(attachedArtifacts);
+                                        list.remove(next);
+                                        try {
+                                            MethodUtils.invokeMethod(project, true, "setAttachedArtifacts", list);
+                                        } catch (NoSuchMethodException | IllegalAccessException
+                                                | InvocationTargetException ignored) {
+                                            log.warn("The attached artifact " + classifier
+                                                    + " is not present in the baseline but could not be removed!");
+                                        }
+                                    }
                                     break;
                                 }
                             }

